packages/errors/src/json-rpc-error.ts

```diff
@@ -0,0 +1,133 @@
+                        parsed: unknown;
+                        program: string;
+                        program_id: string;
+                        stack_height?: number;
+                    }
+                  | {
+                        // PartiallyDecoded
+                        accounts: string[];
+                        data: string;
+                        program_id: string;
+                        stack_height?: number;
+                    }
+              )[];
+          }[]
+        | null;
+    logs: string[] | null;
+    returnData: {
+        data: [string, 'base64'];
+        programId: string;
+    } | null;
+    unitsConsumed: number | null;
+}
+
+export function getSolanaErrorFromJsonRpcError({ code, data, message }: RpcErrorResponse): SolanaError {
+    let out: SolanaError;
+    if (code === SOLANA_ERROR__JSON_RPC__SERVER_ERROR_SEND_TRANSACTION_PREFLIGHT_FAILURE) {
+        const { err, ...preflightErrorContext } = data as RpcSimulateTransactionResult;
+        const causeObject = err ? { cause: getSolanaErrorFromTransactionError(err) } : null;
+        out = new SolanaError(SOLANA_ERROR__JSON_RPC__SERVER_ERROR_SEND_TRANSACTION_PREFLIGHT_FAILURE, {
+            ...preflightErrorContext,
+            cause: causeObject,
+        });
+    } else {
+        let errorContext;
+        switch (code) {
+            case SOLANA_ERROR__JSON_RPC__INTERNAL_ERROR:
+            case SOLANA_ERROR__JSON_RPC__INVALID_PARAMS:
+            case SOLANA_ERROR__JSON_RPC__INVALID_REQUEST:
+            case SOLANA_ERROR__JSON_RPC__METHOD_NOT_FOUND:
+            case SOLANA_ERROR__JSON_RPC__PARSE_ERROR:
+            case SOLANA_ERROR__JSON_RPC__SCAN_ERROR:
+            case SOLANA_ERROR__JSON_RPC__SERVER_ERROR_BLOCK_CLEANED_UP:
+            case SOLANA_ERROR__JSON_RPC__SERVER_ERROR_BLOCK_NOT_AVAILABLE:
+            case SOLANA_ERROR__JSON_RPC__SERVER_ERROR_BLOCK_STATUS_NOT_AVAILABLE_YET:
+            case SOLANA_ERROR__JSON_RPC__SERVER_ERROR_KEY_EXCLUDED_FROM_SECONDARY_INDEX:
+            case SOLANA_ERROR__JSON_RPC__SERVER_ERROR_LONG_TERM_STORAGE_SLOT_SKIPPED:
+            case SOLANA_ERROR__JSON_RPC__SERVER_ERROR_SLOT_SKIPPED:
+            case SOLANA_ERROR__JSON_RPC__SERVER_ERROR_TRANSACTION_PRECOMPILE_VERIFICATION_FAILURE:
+            case SOLANA_ERROR__JSON_RPC__SERVER_ERROR_UNSUPPORTED_TRANSACTION_VERSION:
+                // The server supplies no structured data, but rather a pre-formatted message. Put
+                // the server message in `context` so as not to completely lose the data. The long
+                // term fix for this is to add data to the server responses and modify the
+                // messages in `@solana/errors` to be actual format strings.
+                errorContext = { __serverMessage: message };
+                break;
+            default:
+                if (typeof data === 'object' && !Array.isArray(data)) {
+                    errorContext = data;
+                }
```

Comment:
This protects against things like `data: 'abc'` turning into `{0: 'a', 1: 'b', 2: 'c'}`